sql/src/main/java/io/crate/protocols/postgres/Portal.java

```diff
@@ -59,7 +62,10 @@ Portal bind(String statementName,
 
     void execute(ResultReceiver resultReceiver, int maxRows);
 
-    CompletableFuture<?> sync(Planner planner, JobsLogs jobsLogs);
+    /**
+     * @return True if interrupted, false or null otherwise.
```

Comment:
Where is this `True == interrupt` coming from?

Reply:
https://github.com/crate/crate/blob/master/sql/src/main/java/io/crate/protocols/postgres/PostgresWireProtocol.java#L268

https://github.com/crate/crate/blob/master/sql/src/main/java/io/crate/action/sql/BaseResultReceiver.java#L46

I would like to make it clear what the future returns.

Reply:
I wonder if we should go the other way around and convert the interrupt into an exception. So far we swallowed it anyway, so maybe that's a valid thing to do and we can ignore it?

Or we could introduce an `enum` instead. Going for a `boolean` with `true` as a `not good` value doesn't seem very intuitive to me.

Reply:
It's certainly not intuitive, although having boolean is more intuitive to me than Object :) Enum sounds better.

I guess we could get rid altogether of the return type and just use Void. It doesn't seem like the interrupt is ever used other than deciding whether to not send the ReadyToQuery in case of an error.

Reply:
I would like to fix the return type. It is a major cause of confusion because it is hard to figure out what is returned. So if we don't have a return type, I'd use `Void`. This is standard practice in the Java API, even inside the `CompletableFuture` class.

Maybe I don't understand the type cast implications?

Reply:
Using `CompletableFuture<?>` (or really any `T<?>`) is a common way to indicate
that there is a type checked inner type, but at this point it is irrelevant
what it is because it's not used.

Absence of information is also information.

It has the advantage over `CompletableFuture<Void>` that one can simply pass on
an existing future that has a different inner type, making it unnecessary to
add a `T -> null` conversion. (We also use the same pattern in `loadNextBatch`
of the `BatchIterator`


In any case, I'd be fine with changing it to use `Void`, but I don't think it's
an improvement. And I'm against changing it to `CompletableFuture<Boolean>`,
especially since in the current form in this PR the value is not used at all.
It's adding information indicating that it is important, but it's not.

Reply:
>And I'm against changing it to CompletableFuture<Boolean>,
especially since in the current form in this PR the value is not used at all.
It's adding information indicating that it is important, but it's not.

It is used to check for the interrupted flag. Otherwise the value is hidden by type Object which makes reasoning about the return type hard.

>In any case, I'd be fine with changing it to use Void, but I don't think it's
an improvement. 

Getting rid of the return type and using exceptions together with a Void type is an improvement because people reading the code can make sense of it, instead of deducting the return type from the use of the class in the code.

>Absence of information is also information.

Yes, it's less entropy because there is only one possible outcome. `?` means high entropy => a lot of of possible outcomes.

>It has the advantage over CompletableFuture<Void> that one can simply pass on
an existing future that has a different inner type, making it unnecessary to
add a T -> null conversion. (We also use the same pattern in loadNextBatch
of the BatchIterator

That's not an advantage because it creates a fuzzy contract. Also using it somewhere else already is not a reason to do it here as well.

Reply:
> It is used to check for the interrupted flag. Otherwise the value is hidden by type Object which makes reasoning about the return type hard.

Currently the components calling `sync` do not use the value produced and as far as I can tell that's not changing in this PR. In fact in the places where we could use the explicitly typed return values it's not being used, see https://github.com/crate/crate/pull/7252/files#diff-2acb0e9d2f27ff3488767198223b9ea4R648

> Yes, it's less entropy because there is only one possible outcome. ? means high entropy => a lot of of possible outcomes.

Depends on how you see it. If you see `<?>` as equivalent to `<Object>`, maybe. How it's commonly used it means that some internal components have a concrete known type, and that it's passed through to other components which are not interested in the type.

Anyway, if @seut or @matriv agree that `Void` is better than fine. For me it means we've to waste a couple of extra cycles to convert whatever the inner futures are producing to `Void / null` for a questionable readability improvement.

Reply:
The important outcome of our discussion is that we want to throw an `InterruptedException` and not rely on the boolean result of the future. 

So I'm ok with keeping the ? and documenting the non-existent return type.

Reply:
I don't have a strong opinion here, I think both points are valid. Keeping the ? and add some comment sounds like a good compromise to me.

Reply:
I implemented the exception handling here: e715cf2ff8d6ac4ad9bec3eb677429a3602e5672